at-angular.ts

```diff
@@ -107,8 +144,43 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export interface IDecoratorAnnotation {
+        (moduleName: string, targetProvider: string): IClassAnnotationDecorator;
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
```

Comment:
Maybe it should be generic (parametrized) function:

I think it should finally look like:

```typescript
    export function decorator<A, B>(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {

        interface IDecoratorStatic<D> extends Function {
            new (...args: any[]): D;
        }

        return (targetClass: IDecoratorStatic<B>): void => {

            angular
                .module(moduleName)
                .config([
                    '$provide',
                    function($provide: angular.auto.IProvideService): void {

                        delegation.$inject = ['$delegate', '$injector'];

                        function delegation(
                            $delegate: A,
                            $injector: angular.auto.IInjectorService
                        ): A & B {

                            const instance: B = $injector.instantiate<B>(targetClass, {
                                $delegate: $delegate
                            });

                            return angular.extend($delegate, targetClass.prototype, instance);
                        }

                        $provide.decorator(targetProvider, delegation);
                    }]);

        };

    }
```